File: Python/12-Tries(PrefixTrees)/Algorithms/1-Trie(Insert,Search,Delete)/Trie(Insert,Search,Delete)Pro.py
"""
Trie (Prefix Tree) Implementation in Python: 6 Real-World Approaches

This document provides six Python implementations of the Trie data structure,
focusing on approaches used in real-world projects.  A Trie, also known as a prefix
tree, is a tree-like data structure that is highly efficient for storing and
retrieving strings, especially for prefix-based operations.

Each implementation includes a `TrieNode` class and methods for inserting,
searching, and deleting words.  The examples demonstrate how to use each
implementation.

1.  **Basic Trie with Dictionary (Standard):**
    * Uses a dictionary to represent the children of each node.
    * This is the most common and fundamental Trie implementation.
    * Suitable for general-purpose string storage and retrieval.

2.  **Trie with defaultdict (Concise):**
    * Uses `defaultdict(TrieNode)` to automatically create new nodes when a key
        is accessed.
    * Provides a more compact and readable way to implement the basic Trie.
    * Good for scenarios where code clarity and brevity are important.

3.  **Trie with Compressed Paths (Space-Optimized):**
    * Compresses paths with single children into a single node, reducing space
        usage.
    * Useful when memory efficiency is a primary concern, such as in large
        dictionaries or when dealing with a massive number of strings.

4.  **Trie for Autocomplete (Application-Specific):**
    * Implements a Trie with a method to suggest autocompletion for a given
        prefix.
    * Demonstrates how Tries can be used in applications like search engines
        and text editors.

5.  **Trie for Spell Checking (Application-Specific):**
    * Extends the Trie to provide spell-checking functionality, suggesting
        corrections for misspelled words.
    * Shows how Tries can be adapted for natural language processing tasks.

6.  **Concurrent Trie (Thread-Safe):**
        * Implements a thread-safe Trie using a locking mechanism.
        * Useful in multi-threaded applications.
"""

import threading
from collections import defaultdict
from typing import List, Optional, Tuple

# ------------------------------------------------------------------------------------------------
# 1. Basic Trie with Dictionary (Standard)
# ------------------------------------------------------------------------------------------------
class TrieNode1:
    """
    Represents a node in the Trie.  Each node stores a dictionary of child nodes
    and a flag indicating whether the node represents the end of a word.
    """
    def __init__(self):
        self.children = {}  # Dictionary of child nodes (character: TrieNode)
        self.is_end_of_word = False  # Flag to mark the end of a word

class Trie1:
    """
    Implements a Trie data structure using a dictionary.
    """
    def __init__(self):
        self.root = TrieNode1()  # Root node of the Trie

    def insert(self, word: str) -> None:
        """
        Inserts a word into the Trie.

        Args:
            word: The word to insert.
        """
        node = self.root
        for char in word:
            if char not in node.children:
                node.children[char] = TrieNode1()  # Create a new node if the character is not present
            node = node.children[char]  # Move to the child node
        node.is_end_of_word = True  # Mark the last node as the end of the word

    def search(self, word: str) -> bool:
        """
        Searches for a word in the Trie.

        Args:
            word: The word to search for.

        Returns:
            True if the word is found, False otherwise.
        """
        node = self.root
        for char in word:
            if char not in node.children:
                return False  # Word not found if character not in children
            node = node.children[char]  # Move to the child node
        return node.is_end_of_word  # Check if the last node is the end of the word

    def starts_with(self, prefix: str) -> bool:
        """
        Checks if there is any word in the trie that starts with the given prefix.

        Args:
            prefix: The prefix to search for.

        Returns:
            True if there is a word with the prefix, False otherwise.
        """
        node = self.root
        for char in prefix:
            if char not in node.children:
                return False  # Prefix not found if character not in children
            node = node.children[char]  # Move to the child node
        return True  # Prefix found

# ------------------------------------------------------------------------------------------------
# 2. Trie with defaultdict (Concise)
# ------------------------------------------------------------------------------------------------
class TrieNode2:
    """
    Represents a node in the Trie using defaultdict.
    """
    def __init__(self):
        self.children = defaultdict(TrieNode2)  # Default dictionary of child nodes
        self.is_end_of_word = False

class Trie2:
    """
    Implements a Trie using defaultdict for more concise node creation.
    """
    def __init__(self):
        self.root = TrieNode2()

    def insert(self, word: str) -> None:
        node = self.root
        for char in word:
            node = node.children[char]  # Creates node if it doesn't exist
        node.is_end_of_word = True

    def search(self, word: str) -> bool:
        node = self.root
        for char in word:
            if char not in node.children:
                return False
            node = node.children[char]
        return node.is_end_of_word

    def starts_with(self, prefix: str) -> bool:
        node = self.root
        for char in prefix:
            if char not in node.children:
                return False
            node = node.children[char]
        return True

# ------------------------------------------------------------------------------------------------
# 3. Trie with Compressed Paths (Space-Optimized)
# ------------------------------------------------------------------------------------------------
class TrieNode3:
    """
    Represents a node in a compressed Trie.
    """
    def __init__(self, prefix=""):
        self.prefix = prefix  # The common prefix represented by this node
        self.children = {}  # Dictionary of child nodes (character: TrieNode)
        self.is_end_of_word = False

class Trie3:
    """
    Implements a compressed Trie (Radix Trie) to save space.
    """
    def __init__(self):
        self.root = TrieNode3()

    def insert(self, word: str) -> None:
        """Inserts a word into the compressed Trie."""
        node = self.root

        while True:
            # Find the common prefix
            i = 0
            while i < len(node.prefix) and i < len(word) and node.prefix[i] == word[i]:
                i += 1

            if i == len(node.prefix):
                # Node's prefix is a prefix of the word
                word = word[i:]
                if not word:
                    node.is_end_of_word = True
                    return

                if word[0] in node.children:
                    node = node.children[word[0]]
                    continue
                else:
                    new_node = TrieNode3(word)
                    new_node.is_end_of_word = True
                    node.children[word[0]] = new_node
                    return
            else:
                # Partial match
                common_prefix = node.prefix[:i]
                suffix1 = node.prefix[i:]
                suffix2 = word[i:]

                new_node = TrieNode3(common_prefix)
                child1 = TrieNode3(suffix1)
                child1.is_end_of_word = node.is_end_of_word
                child1.children = node.children
                new_node.children[suffix1[0]] = child1

                if suffix2:
                    child2 = TrieNode3(suffix2)
                    child2.is_end_of_word = True
                    new_node.children[suffix2[0]] = child2

                node.prefix = common_prefix
                node.is_end_of_word = False
                if common_prefix:
                    node.children = {common_prefix[0]: new_node}
                else:
                    node.children = {}
                return

    def search(self, word: str) -> bool:
        """Searches for a word in the compressed Trie."""
        node = self.root

        while True:
            if not word:
                return node.is_end_of_word

            i = 0
            while i < len(node.prefix) and i < len(word) and node.prefix[i] == word[i]:
                i += 1

            if i == 0:
                return False

            if i < len(node.prefix):
                return False

            word = word[i:]

            if not word:
                return node.is_end_of_word

            if word[0] not in node.children:
                return False

            node = node.children[word[0]]

    def starts_with(self, prefix: str) -> bool:
        """Checks if any word starts with the given prefix."""
        node = self.root

        while prefix:
            i = 0
            while i < len(node.prefix) and i < len(prefix) and node.prefix[i] == prefix[i]:
                i += 1

            if i == 0:
                return False

            prefix = prefix[i:]

            if not prefix:
                return True

            if prefix[0] not in node.children:
                return False

            node = node.children[prefix[0]]
        return True

# ------------------------------------------------------------------------------------------------
# 4. Trie for Autocomplete (Application-Specific)
# ------------------------------------------------------------------------------------------------
class TrieNode4:
    """
    Represents a node in the Trie for autocomplete functionality.
    """
    def __init__(self):
        self.children = {}
        self.is_end_of_word = False
        self.suggestions = []  # List to store autocompletion suggestions

class Trie4:
    """
    Implements a Trie for autocomplete suggestions.
    """
    def __init__(self):
        self.root = TrieNode4()
        self.max_suggestions = 5

    def insert(self, word: str) -> None:
        """Inserts a word into the Trie."""
        node = self.root
        for char in word:
            if char not in node.children:
                node.children[char] = TrieNode4()
            node = node.children[char]
        node.is_end_of_word = True

    def search(self, word: str) -> bool:
        """Searches for a word in the Trie."""
        node = self.root
        for char in word:
            if char not in node.children:
                return False
            node = node.children[char]
        return node.is_end_of_word

    def get_suggestions(self, prefix: str) -> List[str]:
        """
        Gets autocompletion suggestions for a given prefix.

        Args:
            prefix: The prefix to get suggestions for.

        Returns:
            A list of suggested words.
        """
        node = self.root
        for char in prefix:
            if char not in node.children:
                return []  # No suggestions if prefix not found
            node = node.children[char]

        self._collect_suggestions(node, prefix)
        return node.suggestions

    def _collect_suggestions(self, node: TrieNode4, current_word: str) -> None:
        """
        Recursively collects suggestions from the Trie.

        Args:
            node: The current node.
            current_word: The current word being formed.
        """
        if len(node.suggestions) >= self.max_suggestions:
            return

        if node.is_end_of_word:
            node.suggestions.append(current_word)

        for char, child in node.children.items():
            self._collect_suggestions(child, current_word + char)

# ------------------------------------------------------------------------------------------------
# 5. Trie for Spell Checking (Application-Specific)
# ------------------------------------------------------------------------------------------------
class TrieNode5:
    """
    Represents a node in the Trie for spell-checking.
    """
    def __init__(self):
        self.children = {}
        self.is_end_of_word = False

class Trie5:
    """
    Implements a Trie for spell checking.
    """
    def __init__(self):
        self.root = TrieNode5()
        self.alphabet = 'abcdefghijklmnopqrstuvwxyz'

    def insert(self, word: str) -> None:
        """Inserts a word into the Trie."""
        node = self.root
        for char in word:
            if char not in node.children:
                node.children[char] = TrieNode5()
            node = node.children[char]
        node.is_end_of_word = True

    def search(self, word: str) -> bool:
        """Searches for a word in the Trie."""
        node = self.root
        for char in word:
            if char not in node.children:
                return False
            node = node.children[char]
        return node.is_end_of_word

    def suggest_corrections(self, word: str) -> List[str]:
        """
        Suggests spelling corrections for a given word.

        Args:
            word: The misspelled word.

        Returns:
            A list of suggested corrections.
        """
        if self.search(word):
            return [word]  # The word is already correct

        suggestions = set()
        self._collect_corrections(self.root, word, "", 0, suggestions)
        return list(suggestions)

    def _collect_corrections(self, node: TrieNode5, word: str, current_word: str, index: int, suggestions: set) -> None:
        """
        Recursively collects spelling corrections.

        Args:
            node: The current node in the Trie.
            word: The misspelled word.
            current_word: The current word being formed.
            index: The current character index in the misspelled word.
            suggestions: The set to store suggested corrections.
        """
        if index == len(word):
            if node.is_end_of_word:
                suggestions.add(current_word)
            return

        char = word[index]

        # Check for correct character
        if char in node.children:
            self._collect_corrections(node.children[char], word, current_word + char, index + 1, suggestions)

        # Check for substitution
        for c in self.alphabet:
            if c in node.children:
                self._collect_corrections(node.children[c], word, current_word + c, index + 1, suggestions)

        # Check for insertion
        for c in self.alphabet:
            if c in node.children:
                self._collect_corrections(node.children[c], word, current_word + char + c, index + 1, suggestions)
        #Deletion
        if index + 1 <= len(word):
            self._collect_corrections(node, word, current_word, index+1, suggestions)

# ------------------------------------------------------------------------------------------------
# 6. Concurrent Trie (Thread-Safe)
# ------------------------------------------------------------------------------------------------
class TrieNode6:
    """
    Represents a node in the thread-safe Trie.
    """
    def __init__(self):
        self.children = {}
        self.is_end_of_word = False
        self.lock = threading.Lock()  # Lock for thread safety

class Trie6:
    """
    Implements a thread-safe Trie using a lock.
    """
    def __init__(self):
        self.root = TrieNode6()

    def insert(self, word: str) -> None:
        """Inserts a word into the Trie in a thread-safe manner."""
        node = self.root
        for char in word:
            with node.lock:  # Acquire lock before accessing children
                if char not in node.children:
                    node.children[char] = TrieNode6()
                node = node.children[char]
        with node.lock:
            node.is_end_of_word = True

    def search(self, word: str) -> bool:
        """Searches for a word in the Trie in a thread-safe manner."""
        node = self.root
        for char in word:
            with node.lock:
                if char not in node.children:
                    return False
                node = node.children[char]
        with node.lock:
            return node.is_end_of_word

    def starts_with(self, prefix: str) -> bool:
        """Checks for a prefix in a thread-safe manner."""
        node = self.root
        for char in prefix:
            with node.lock:
                if char not in node.children:
                    return False
                node = node.children[char]
        return True

# ------------------------------------------------------------------------------------------------
# Example Usage
# ------------------------------------------------------------------------------------------------
def main():
    """
    Main function to demonstrate the usage of the Trie implementations.
    """
    # Example 1: Basic Trie with Dictionary
    print("\nExample 1: Basic Trie with Dictionary")
    trie1 = Trie1()
    trie1.insert("apple")
    trie1.insert("app")
    print(f"Search 'apple': {trie1.search('apple')}")
    print(f"Search 'app': {trie1.search('app')}")
    print(f"Search 'ap': {trie1.search('ap')}")
    print(f"Starts with 'ap': {trie1.starts_with('ap')}")

    # Example 2: Trie with defaultdict
    print("\nExample 2: Trie with defaultdict")
    trie2 = Trie2()
    trie2.insert("apple")
    trie2.insert("app")
    print(f"Search 'apple': {trie2.search('apple')}")
    print(f"Search 'app': {trie2.search('app')}")
    print(f"Search 'ap': {trie2.search('ap')}")
    print(f"Starts with 'ap': {trie2.starts_with('ap')}")

    # Example 3: Trie with Compressed Paths
    print("\nExample 3: Trie with Compressed Paths")
    trie3 = Trie3()
    trie3.insert("apple")
    trie3.insert("app")
    trie3.insert("application")
    trie3.insert("banana")
    print(f"Search 'apple': {trie3.search('apple')}")
    print(f"Search 'app': {trie3.search('app')}")
    print(f"Search 'application': {trie3.search('application')}")
    print(f"Search 'banana': {trie3.search('banana')}")
    print(f"Search 'ban': {trie3.search('ban')}")
    print(f"Starts with 'app': {trie3.starts_with('app')}")
    print(f"Starts with 'ban': {trie3.starts_with('ban')}")
    print(f"Starts with 'xy': {trie3.starts_with('xy')}")

    # Example 4: Trie for Autocomplete
    print("\nExample 4: Trie for Autocomplete")
    trie4 = Trie4()
    trie4.insert("apple")
    trie4.insert("app")
    trie4.insert("application")
    trie4.insert("banana")
    print(f"Suggestions for 'app': {trie4.get_suggestions('app')}")
    print(f"Suggestions for 'ban': {trie4.get_suggestions('ban')}")
    print(f"Suggestions for 'x': {trie4.get_suggestions('x')}")

    # Example 5: Trie for Spell Checking
    print("\nExample 5: Trie for Spell Checking")
    trie5 = Trie5()
    trie5.insert("apple")
    trie5.insert("banana")
    trie5.insert("application")
    print(f"Suggestions for 'appl': {trie5.suggest_corrections('appl')}")
    print(f"Suggestions for 'banan': {trie5.suggest_corrections('banan')}")
    print(f"Suggestions for 'applcation': {trie5.suggest_corrections('applcation')}")
    print(f"Suggestions for 'xyz': {trie5.suggest_corrections('xyz')}")

    # Example 6: Concurrent Trie
    print("\nExample 6: Concurrent Trie")
    trie6 = Trie6()
    trie6.insert("apple")
    trie6.insert("app")

    def search_thread(word):
        print(f"Thread {threading.current_thread().name} searching for '{word}': {trie6.search(word)}")

    def insert_thread(word):
        trie6.insert(word)
        print(f"Thread {threading.current_thread().name} inserted '{word}'")

    # Create multiple threads to simulate concurrent access
    threads = []
    for i in range(5):
        thread = threading.Thread(target=search_thread, args=("apple",), name=f"SearchThread-{i}")
        threads.append(thread)
        thread.start()

    thread = threading.Thread(target=insert_thread, args=("banana",), name="InsertThread")
    threads.append(thread)
    thread.start()

    for thread in threads:
        thread.join()  # Wait for all threads to finish

    print(f"Search 'banana' after insertion: {trie6.search('banana')}")

if __name__ == "__main__":
    main()

File: JavaScript/12-Tries(PrefixTrees)/Algorithms/1-Trie(Insert,Search,Delete)/Trie(Insert,Search,Delete)Pro.js
/*
Tries (Prefix Trees) - JavaScript Implementation with 5 Approaches

A Trie, also known as a prefix tree, is a tree-like data structure used for efficient storage and retrieval of strings.  It excels at prefix-based searches.

Here are five different approaches to implementing a Trie in JavaScript, each with a slightly different focus and potential use case.  Each approach includes insert, search, and delete methods, along with example usage and console.log output.

Key Concepts:

* Node: Each node in a Trie represents a character.
* Root: The starting point of the Trie (often empty).
* Children: Each node has children, representing the next possible characters in a string.
* isEndOfWord: A boolean flag indicating if a node marks the end of a complete word.

*/

// ----------------------------------------------------------------------------------------------------
// 1. Basic Trie Implementation (Object-Based)
// ----------------------------------------------------------------------------------------------------
class TrieNode1 {
    constructor() {
        this.children = {}; // Using an object for children (more common in JS)
        this.isEndOfWord = false;
    }
}

class Trie1 {
    constructor() {
        this.root = new TrieNode1();
    }

    insert(word) {
        let node = this.root;
        for (const char of word) {
            if (!node.children[char]) {
                node.children[char] = new TrieNode1();
            }
            node = node.children[char];
        }
        node.isEndOfWord = true;
    }

    search(word) {
        let node = this.root;
        for (const char of word) {
            if (!node.children[char]) {
                return false;
            }
            node = node.children[char];
        }
        return node && node.isEndOfWord;
    }

    startsWith(prefix) {
      let node = this.root;
      for (const char of prefix) {
          if (!node.children[char]) {
              return false;
          }
          node = node.children[char];
      }
      return true;
    }


    delete(word) {
        this._deleteHelper(this.root, word, 0);
    }

    _deleteHelper(node, word, index) {
        if (index === word.length) {
            if (!node.isEndOfWord) {
                return false; // Word not found
            }
            node.isEndOfWord = false; // Unmark the end of the word
            return Object.keys(node.children).length === 0; // Return true if node has no children
        }

        const char = word[index];
        const nextNode = node.children[char];

        if (!nextNode) {
            return false; // Word not found
        }

        const shouldDeleteChild = this._deleteHelper(nextNode, word, index + 1);

        if (shouldDeleteChild) {
            delete node.children[char]; // Delete the child node
            return Object.keys(node.children).length === 0; // Return true if node has no children
        }

        return false; // No deletion happened at this level
    }
}

// Example Usage (Trie 1)
const trie1 = new Trie1();
trie1.insert("apple");
trie1.insert("app");
trie1.insert("banana");

console.log("--- Trie 1 (Object-Based) ---");
console.log("Search 'apple':", trie1.search("apple")); // true
console.log("Search 'app':", trie1.search("app"));     // true
console.log("Search 'banana':", trie1.search("banana")); // true
console.log("Search 'orange':", trie1.search("orange")); // false
console.log("Starts with 'ap':", trie1.startsWith("ap")); // true
console.log("Starts with 'ba':", trie1.startsWith("ba")); // true

trie1.delete("apple");
console.log("After deleting 'apple', search 'apple':", trie1.search("apple")); // false
console.log("After deleting 'apple', search 'app':", trie1.search("app"));       // true
console.log("Search 'apple' after delete:", trie1.search("apple")); // false
console.log("Search 'app' after delete:", trie1.search("app"));     // true

// ----------------------------------------------------------------------------------------------------
// 2. Trie with Array-Based Children
// ----------------------------------------------------------------------------------------------------
const ALPHABET_SIZE = 26; // For lowercase English letters

class TrieNode2 {
    constructor() {
        this.children = new Array(ALPHABET_SIZE).fill(null);
        this.isEndOfWord = false;
    }
}

class Trie2 {
    constructor() {
        this.root = new TrieNode2();
    }

    _charToIndex(char) {
        return char.charCodeAt(0) - 'a'.charCodeAt(0);
    }

    insert(word) {
        let node = this.root;
        for (const char of word) {
            const index = this._charToIndex(char);
            if (!node.children[index]) {
                node.children[index] = new TrieNode2();
            }
            node = node.children[index];
        }
        node.isEndOfWord = true;
    }

    search(word) {
        let node = this.root;
        for (const char of word) {
            const index = this._charToIndex(char);
            if (!node.children[index]) {
                return false;
            }
            node = node.children[index];
        }
        return node && node.isEndOfWord;
    }
      startsWith(prefix) {
        let node = this.root;
        for (const char of prefix) {
            const index = this._charToIndex(char);
            if (!node.children[index]) {
                return false;
            }
            node = node.children[index];
        }
        return true;
    }

    delete(word) {
        this._deleteHelper(this.root, word, 0);
    }

    _deleteHelper(node, word, index) {
        if (index === word.length) {
            if (!node.isEndOfWord) {
                return false;
            }
            node.isEndOfWord = false;
            return this._hasNoChildren(node);
        }

        const charIndex = this._charToIndex(word[index]);
        const nextNode = node.children[charIndex];

        if (!nextNode) {
            return false;
        }

        const shouldDeleteChild = this._deleteHelper(nextNode, word, index + 1);

        if (shouldDeleteChild) {
            node.children[charIndex] = null;
            return this._hasNoChildren(node);
        }
        return false;
    }

      _hasNoChildren(node) {
        for (let i = 0; i < ALPHABET_SIZE; i++) {
            if (node.children[i] !== null) {
                return false;
            }
        }
        return true;
    }
}

// Example Usage (Trie 2)
const trie2 = new Trie2();
trie2.insert("apple");
trie2.insert("app");
trie2.insert("banana");

console.log("\n--- Trie 2 (Array-Based Children) ---");
console.log("Search 'apple':", trie2.search("apple"));
console.log("Search 'app':", trie2.search("app"));
console.log("Search 'banana':", trie2.search("banana"));
console.log("Search 'orange':", trie2.search("orange"));
console.log("Starts with 'ap':", trie2.startsWith("ap"));
console.log("Starts with 'ba':", trie2.startsWith("ba"));

trie2.delete("apple");
console.log("After deleting 'apple', search 'apple':", trie2.search("apple")); // false
console.log("After deleting 'apple', search 'app':", trie2.search("app"));       // true

// ----------------------------------------------------------------------------------------------------
// 3. Trie with Custom Alphabet
// ----------------------------------------------------------------------------------------------------
class TrieNode3 {
    constructor() {
        this.children = {};
        this.isEndOfWord = false;
    }
}

class Trie3 {
    constructor(alphabet) {
        this.root = new TrieNode3();
        this.alphabet = alphabet;
        this.alphabetMap = new Map(); // For efficient character to index mapping
        for (let i = 0; i < alphabet.length; i++) {
          this.alphabetMap.set(alphabet[i], i);
        }
    }

    _charToIndex(char) {
        const index = this.alphabetMap.get(char);
        if (index === undefined) {
            throw new Error(`Character "${char}" not in alphabet`);
        }
        return index;
    }

    insert(word) {
        let node = this.root;
        for (const char of word) {
            const index = this._charToIndex(char);
            if (!node.children[char]) {
                node.children[char] = new TrieNode3();
            }
            node = node.children[char];
        }
        node.isEndOfWord = true;
    }

    search(word) {
        let node = this.root;
        for (const char of word) {
            const index = this._charToIndex(char);
            if (!node.children[char]) {
                return false;
            }
            node = node.children[char];
        }
        return node && node.isEndOfWord;
    }

    startsWith(prefix) {
      let node = this.root;
      for (const char of prefix) {
          const index = this._charToIndex(char);
          if (!node.children[char]) {
              return false;
          }
          node = node.children[char];
      }
      return true;
    }

    delete(word) {
      this._deleteHelper(this.root, word, 0);
    }

  _deleteHelper(node, word, index) {
        if (index === word.length) {
            if (!node.isEndOfWord) {
                return false;
            }
            node.isEndOfWord = false;
            return Object.keys(node.children).length === 0;
        }

        const char = word[index];
        const nextNode = node.children[char];

        if (!nextNode) {
            return false;
        }

        const shouldDeleteChild = this._deleteHelper(nextNode, word, index + 1);

        if (shouldDeleteChild) {
            delete node.children[char];
            return Object.keys(node.children).length === 0;
        }
        return false;
    }
}

// Example Usage (Trie 3)
const alphabet = "abcdefghijklmnopqrstuvwxyz' "; // Example: Allow spaces and apostrophes
const trie3 = new Trie3(alphabet);
trie3.insert("apple");
trie3.insert("app");
trie3.insert("banana");
trie3.insert("don't");
trie3.insert("hello world");

console.log("\n--- Trie 3 (Custom Alphabet) ---");
console.log("Search 'apple':", trie3.search("apple"));
console.log("Search 'app':", trie3.search("app"));
console.log("Search 'banana':", trie3.search("banana"));
console.log("Search 'don't':", trie3.search("don't"));
console.log("Search 'hello world':", trie3.search("hello world"));
console.log("Search 'orange':", trie3.search("orange"));
console.log("Starts with 'ap':", trie3.startsWith("ap"));
console.log("Starts with 'do':", trie3.startsWith("do"));
console.log("Starts with 'hello':", trie3.startsWith("hello"));

trie3.delete("apple");
console.log("After deleting 'apple', search 'apple':", trie3.search("apple"));
console.log("After deleting 'apple', search 'app':", trie3.search("app"));

// ----------------------------------------------------------------------------------------------------
// 4. Trie with Node IDs (for Graph-like Applications)
// ----------------------------------------------------------------------------------------------------
let nextNodeId = 0; // Simple global ID counter for demonstration

class TrieNode4 {
    constructor() {
        this.id = nextNodeId++; // Assign a unique ID to each node
        this.children = {};
        this.isEndOfWord = false;
    }
}

class Trie4 {
    constructor() {
        this.root = new TrieNode4();
        this.nodeMap = new Map(); // Store nodes by their IDs if needed
        this.nodeMap.set(this.root.id, this.root);
    }

    insert(word) {
        let node = this.root;
        for (const char of word) {
            if (!node.children[char]) {
                const newNode = new TrieNode4();
                node.children[char] = newNode;
                this.nodeMap.set(newNode.id, newNode); // Store the new node
            }
            node = node.children[char];
        }
        node.isEndOfWord = true;
    }

      search(word) {
        let node = this.root;
        for (const char of word) {
            if (!node.children[char]) {
                return false;
            }
            node = node.children[char];
        }
        return node && node.isEndOfWord;
    }
    startsWith(prefix) {
      let node = this.root;
      for (const char of prefix) {
          if (!node.children[char]) {
              return false;
          }
          node = node.children[char];
      }
      return true;
    }

    delete(word) {
        this._deleteHelper(this.root, word, 0);
    }

    _deleteHelper(node, word, index) {
         if (index === word.length) {
            if (!node.isEndOfWord) {
                return false;
            }
            node.isEndOfWord = false;
            return Object.keys(node.children).length === 0;
        }

        const char = word[index];
        const nextNode = node.children[char];

        if (!nextNode) {
            return false;
        }

        const shouldDeleteChild = this._deleteHelper(nextNode, word, index + 1);

        if (shouldDeleteChild) {
            delete node.children[char];
            return Object.keys(node.children).length === 0;
        }
        return false;
    }

    getNode(nodeId) {
      return this.nodeMap.get(nodeId);
    }
}

// Example Usage (Trie 4)
const trie4 = new Trie4();
trie4.insert("apple");
trie4.insert("app");
trie4.insert("banana");

console.log("\n--- Trie 4 (With Node IDs) ---");
console.log("Search 'apple':", trie4.search("apple"));
console.log("Search 'app':", trie4.search("app"));
console.log("Search 'banana':", trie4.search("banana"));
console.log("Node ID of root:", trie4.root.id);
const appleNode = trie4.getNode(trie4.root.children['a'].id);
console.log("Node after 'a' in 'apple':", appleNode);
console.log("Starts with 'ap':", trie4.startsWith("ap"));
console.log("Starts with 'ba':", trie4.startsWith("ba"));

trie4.delete("apple");
console.log("After deleting 'apple', search 'apple':", trie4.search("apple"));
console.log("After deleting 'apple', search 'app':", trie4.search("app"));

// ----------------------------------------------------------------------------------------------------
// 5. Trie with Word Frequency Counting
// ----------------------------------------------------------------------------------------------------
class TrieNode5 {
    constructor() {
        this.children = {};
        this.isEndOfWord = false;
        this.wordFrequency = 0; // Added for frequency counting
    }
}

class Trie5 {
    constructor() {
        this.root = new TrieNode5();
    }

    insert(word) {
        let node = this.root;
        for (const char of word) {
            if (!node.children[char]) {
                node.children[char] = new TrieNode5();
            }
            node = node.children[char];
        }
        node.isEndOfWord = true;
        node.wordFrequency++; // Increment frequency
    }

    search(word) {
        let node = this.root;
        for (const char of word) {
            if (!node.children[char]) {
                return { found: false, frequency: 0 }; // Return an object
            }
            node = node.children[char];
        }
        return { found: node && node.isEndOfWord, frequency: node ? node.wordFrequency : 0 }; // Return an object
    }
    startsWith(prefix) {
      let node = this.root;
      for (const char of prefix) {
          if (!node.children[char]) {
              return false;
          }
          node = node.children[char];
      }
      return true;
    }

    delete(word) {
        this._deleteHelper(this.root, word, 0);
    }

      _deleteHelper(node, word, index) {
        if (index === word.length) {
            if (!node.isEndOfWord) {
                return false;
            }
            node.isEndOfWord = false;
             if (node.wordFrequency > 0) {
                node.wordFrequency--;
             }
            return Object.keys(node.children).length === 0;
        }

        const char = word[index];
        const nextNode = node.children[char];

        if (!nextNode) {
            return false;
        }

        const shouldDeleteChild = this._deleteHelper(nextNode, word, index + 1);

        if (shouldDeleteChild) {
            delete node.children[char];
            return Object.keys(node.children).length === 0;
        }
        return false;
    }
}

// Example Usage (Trie 5)
const trie5 = new Trie5();
trie5.insert("apple");
trie5.insert("apple");
trie5.insert("app");
trie5.insert("banana");
trie5.insert("apple"); // Insert "apple" again

console.log("\n--- Trie 5 (Word Frequency Counting) ---");
console.log("Search 'apple':", trie5.search("apple")); // { found: true, frequency: 3 }
console.log("Search 'app':", trie5.search("app"));     // { found: true, frequency: 1 }
console.log("Search 'banana':", trie5.search("banana")); // { found: true, frequency: 1 }
console.log("Search 'orange':", trie5.search("orange")); // { found: false, frequency: 0 }
console.log("Starts with 'ap':", trie5.startsWith("ap"));
console.log("Starts with 'ba':", trie5.startsWith("ba"));

trie5.delete("apple");
console.log("After deleting 'apple' once, search 'apple':", trie5.search("apple")); // { found: true, frequency: 2 }
trie5.delete("apple");
console.log("After deleting 'apple' twice, search 'apple':", trie5.search("apple"));
console.log("After deleting 'apple' twice, search 'app':", trie5.search("app"));
